crates/solver/src/liquidity/zeroex.rs

```diff
@@ -9,39 +9,287 @@ use {
+        // this call will error if the socket is already closed but that's fine
+        let _ = socket.close(None).await;
+        result
+    }
+
+    /// Indifinitelly reads messages from the websocket and updates the cache.
+    async fn update_cache(&self, socket: &mut Socket) -> Result<(), UpdateError> {
+        while let Some(msg) = socket.next().await {
+            let msg = msg.context("websocket error")?;
+            let text = match msg {
+                Message::Text(text) => text,
+                Message::Close(frame) => {
+                    return Err(UpdateError::SocketClosed(
+                        frame.map(|frame| anyhow!(frame.reason)),
+                    ))
+                }
+                Message::Ping(payload) => {
+                    // send pong message to keep the connection alive
+                    socket
+                        .send(Message::Pong(payload))
+                        .await
+                        .context("ping pong failure")?;
+                    continue;
+                }
+                _ => {
+                    tracing::error!("Received unsupported message {:?}", msg);
+                    return Err(UpdateError::UnsupportedMessage);
+                }
+            };
+
+            let records = serde_json::from_str::<OrdersResponse>(&text)
+                .with_context(|| format!("deserialization error, text received: {}", text))
+                .map_err(UpdateError::DeserializeError)?
+                .payload;
+
+            let mut cache = self.orders.lock().await;
+            for record in records {
+                match record.metadata.state {
+                    OrderState::Added | OrderState::Updated | OrderState::Fillable => {
+                        cache.insert(hex::encode(record.metadata.order_hash.clone()), record);
+                    }
+                    OrderState::Expired => {
+                        cache.remove(&hex::encode(record.metadata.order_hash));
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Main update loop
+    pub async fn update(&self, init_api: Arc<dyn ZeroExApi>, update_api: ZeroExWebsocketApi) {
+        let mut backoff = DEFAULT_BACKOFF;
+        // loop needed for reconnections
+        loop {
+            // initialize cache by fetching all existing orders via http api
+            if let Err(err) = self.init_cache(init_api.clone()).await {
+                tracing::warn!(
+                    "Error initializing 0x cache: {:?}, retrying with backoff {:?} ...",
+                    err,
+                    backoff
+                );
+                tokio::time::sleep(backoff).await;
+                backoff *= 2;
+                continue;
+            }
+
+            backoff = DEFAULT_BACKOFF;
+            // from now on, rely on websocket connection to do incremental updates
+            if let Err(err) = self.connect_and_update_cache(update_api.clone()).await {
```

Comment:
This can lead to errors in some scenarios:
1. start filling cache with slow http API
2. in the mean time some orders change their state (maybe filled or closed)
3. finished filling cache
4. init web socket connection

While we are initializing the cache we might already miss events.

Maybe we can switch the order and first subscribe to the update with the websocket and while that connection is open fetch all the current orders from the HTTP API. If we already received an update from the websocket for a given order we can simply discard it from the HTTP update.